### PR TITLE
Add client side aggregation

### PR DESCRIPTION
**Pull request by andrewqian2001datadog**
### Requirements for Contributing to this repository

* Fill out the template below. Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion.
* The pull request must only fix one issue, or add one feature, at the time.
* The pull request must update the test suite to demonstrate the changed functionality.
* After you create the pull request, all status checks must be pass before a maintainer reviews your contribution. For more details, please see [CONTRIBUTING](/CONTRIBUTING.md). 

### What does this PR do?

[Project description
](https://docs.google.com/document/d/11fPSNMnp5tMLRK16VgXa6FJ7fTpzggKevpiO6MsQ0H0/edit?usp=sharing)
This PR adds simple client side aggregation (aggregating metrics before they are received by the agent) for the python client. This PR combines two previous PRs together ([1](https://github.com/DataDog/datadogpy/pull/841), [2](https://github.com/DataDog/datadogpy/pull/837))

Client side aggregation has already been implemented in [Go](https://github.com/DataDog/datadog-go/pull/139), Java, and the C# dogstatsd clients

### Description of the Change

Periodically aggregates metrics (Count, Gauge, Set) for a given `statsd_aggregation_flush_interval` and then sends the aggregated metrics to intake. A metric will be aggregated if `statsd_disable_aggregating` is `False` and there was another [matching context](https://docs.google.com/document/d/1AoveWKoMQc65hYbvZ27uE7pQ_I5nsGfvv25EcNVPLvQ/edit#heading=h.no4x6fjugsh) sent within the same flush interval.

### Verification Process

Unit tests that verify the aggregator.py class matches the existing behavior in the client side aggregator for the [go client](https://github.com/DataDog/datadog-go/pull/139/files#diff-f554159d70c08124ea27e909de77e6b7162f782cbc12363a0cac8825834a2210).

Unit tests that verify the new metric objects (merics.py) match the existing behavior in the [client side aggregator for Go](https://github.com/DataDog/datadogpy/pull/url).

**Local testing**:
[Confluence doc](https://datadoghq.atlassian.net/wiki/spaces/~71202058eaddc626774b6382b0a98b3e9f4a84/pages/3978855545/Local+testing+for+python+client+side+aggregation+project)
[More Docs](https://dddev.datadoghq.com/notebook/9246140/python-client-side-aggregation-testing)

Bench Mark tests (TODO as a stretch goal)

### Additional Notes

We currently do not allow buffering and aggregation at the same time, if both enabled, aggregation will occur instead of buffering.
### Review checklist (to be filled by reviewers)

- [x] Feature or bug fix MUST have appropriate tests (unit, integration, etc...)
- [ ] PR title must be written as a CHANGELOG entry [(see why)](https://github.com/DataDog/datadogpy/blob/master/CONTRIBUTING.md#pull-request-title)
- [x] Files changes must correspond to the primary purpose of the PR as described in the title (small unrelated changes should have their own PR)
- [x] PR must have one `changelog/` label attached. If applicable it should have the `backward-incompatible` label attached.
- [x] PR should not have `do-not-merge/` label attached.
- [ ] If Applicable, issue must have `kind/` and `severity/` labels attached at least.

### CODE REVIEW

File: datadog/__init__.py [hunk 1]
```diff
@@ -37,7 +37,9 @@ def initialize(
     api_host=None,  # type: Optional[str]
     statsd_host=None,  # type: Optional[str]
     statsd_port=None,  # type: Optional[int]
+    statsd_disable_aggregating=False,  # type: bool
```
Comment (gh123man) [inline datadog/__init__.py:40]: I think at least for now we should default this to disabled. Once we do more testing and wider deployments on more workloads we can consider making it the default :)

File: datadog/dogstatsd/base.py [hunk 2]
```diff
@@ -24,9 +24,13 @@
     # pypy has the same module, but capitalized.
     import Queue as queue  # type: ignore[no-redef]
 
+# pylint: disable=unused-import
```
Comment (gh123man) [inline datadog/dogstatsd/base.py:27]: NIT can we resolve/remove this? If an import is unused, I think we can omit it?

File: datadog/dogstatsd/base.py [hunk 3]
```diff
@@ -45,9 +49,12 @@
...
 
 # Buffering-related values (in seconds)
-DEFAULT_FLUSH_INTERVAL = 0.3
-MIN_FLUSH_INTERVAL = 0.0001
+DEFAULT_BUFFERING_FLUSH_INTERVAL = 0.3
+MIN_BUFFERING_FLUSH_INTERVAL = 0.0001
 
+# Aggregation-related values (in seconds)
+DEFAULT_AGGREGATION_FLUSH_INTERVAL = 2
+MIN_AGGREGATION_FLUSH_INTERVAL = 0.01
```
Comment (gh123man) [inline datadog/dogstatsd/base.py:57]: Maybe this can be the same as `MIN_BUFFERING_FLUSH_INTERVAL`? (we could just keep `MIN_FLUSH_INTERVAL` as it was and use it for both?) 

Unless this new value has some significance. I don't see why the minimum couldn't be much smaller even if impractical.

File: datadog/dogstatsd/base.py [hunk 4]
```diff
@@ -128,31 +138,35 @@ class DogStatsd(object):
...
-        max_buffer_len=0,                       # type: int
-        container_id=None,                      # type: Optional[Text]
-        origin_detection_enabled=True,          # type: bool
-        socket_timeout=0,                       # type: Optional[float]
-        telemetry_socket_timeout=0,             # type: Optional[float]
-        disable_background_sender=True,         # type: bool
-        sender_queue_size=0,                    # type: int
-        sender_queue_timeout=0,                 # type: Optional[float]
-        track_instance=True,                    # type: bool
+        host=DEFAULT_HOST,  # type: Text
```
Comment (gh123man) [inline datadog/dogstatsd/base.py:141]: it looks like the formatting got changed here. 
I'm not sure if a linter was flagging this or if a formatter was misconfigured, but the previous (column aligned) formatting is better IMO. Maybe we can revert this?

File: datadog/dogstatsd/base.py [hunk 5]
```diff
@@ -328,8 +345,9 @@ def __init__(
 
         # Check for deprecated option
         if max_buffer_size is not None:
-            log.warning("The parameter max_buffer_size is now deprecated and is not used anymore")
-
+            log.warning(
```
Comment (gh123man) [inline datadog/dogstatsd/base.py:348]: I think your linter may have [ignored the config](https://github.com/DataDog/datadogpy/blob/master/tox.ini#L49). it looks like this got line truncated, but our line limit is 120 chars. 

Can you re-run the linter suggestions with the config applied? It will remove several unnecessary changes in this diff.

File: datadog/dogstatsd/base.py [hunk 6]
```diff
@@ -423,26 +441,51 @@ def __init__(
...
-        self._send = self._send_to_buffer
         self._disable_buffering = disable_buffering
-        if self._disable_buffering:
-            self._send = self._send_to_server
-            log.debug("Statsd buffering is disabled")
+        self._disable_aggregating = disable_aggregating
 
+        self._flush_interval = flush_interval
+        self._aggregation_flush_interval = aggregation_flush_interval
+        # We make the _buffering_flush_thread and _aggregation_flush_thread a list so that it is a mutable object,
```
Comment (gh123man) [inline datadog/dogstatsd/base.py:453]: is there a reason we cannot continue using `self._flush_thread` in favor of the list thread containers? 
Since aggregation and buffering won't be enabled at the same time (at this point) `self._flush_thread` will only be occupied by one or the other routine.

**Review by gh123man — CHANGES_REQUESTED** [review]
Did a first pass. Left a few comments/questions. 

I think the first thing to address is getting the linter to not change a bunch of unnecessary lines since it makes it hard to determine which lines are functional changes, and which is just formatting. 

Ill do another review once these are addressed :)

File: datadog/dogstatsd/base.py (hunk 2 above)
Comment (andrewqian2001datadog) [reply datadog/dogstatsd/base.py:27]: Its used in the comments and they give me a lint issue when I removed them

File: datadog/dogstatsd/base.py (hunk 6 above)
Comment (andrewqian2001datadog) [reply datadog/dogstatsd/base.py:453]: No I think that works as well, I just thought designing it like this would be more flexible and at the time I wrote the code I dont think I knew whether or not they would be allowed at the same time.

Comment (gh123man) [reply datadog/dogstatsd/base.py:453]: Makes sense. In that case - let's do the simple thing for now and reuse `self. _flush_thread`.

Comment (andrewqian2001datadog) [reply datadog/dogstatsd/base.py:453]: Sounds good

**Review by gh123man — CHANGES_REQUESTED** [review]

File: datadog/dogstatsd/base.py [hunk 7]
```diff
@@ -541,27 +585,27 @@ def _flush_thread_loop(self, flush_interval):
...
         )
 
     # Note: Invocations of this method should be thread-safe
     def _stop_flush_thread(self):
         if not self._flush_thread:
             return
-
         try:
-            self.flush()
+            if self.disable_aggregation:
```
[the remark this answers is not in the document]
Comment (andrewqian2001datadog) [reply datadog/dogstatsd/base.py:595]: Oh this was mb it was supposed to be self._disable_aggregating, thanks!

File: datadog/dogstatsd/base.py [hunk 8]
```diff
@@ -598,7 +642,35 @@ def disable_buffering(self, is_disabled):
...
+                return
+
+            self._disable_aggregating = True
+
+            # If aggregation has been disabled, flush and kill the background thread
+            # otherwise start up the flushing thread and enable aggregation.
+            self._stop_flush_thread()
+            log.debug("Statsd aggregation is disabled")
+
+    def enable_aggregation(self, aggregation_flush_interval=DEFAULT_AGGREGATION_FLUSH_INTERVAL):
```
[the remark this answers is not in the document]
Comment (andrewqian2001datadog) [reply datadog/dogstatsd/base.py:659]: the __init__ function is sometimes not called with any parameters set and just has the default parameters as we don't require the user to call initialize before sending metrics. So I believe we want a way to start aggregation.

This was a convo I had with Vikentiy 

"That’s because we create the global statsd instance with all default parameters ([here](https://github.com/DataDog/datadogpy/blob/add-client-side-aggregation/datadog/dogstatsd/base.py#L1630)) - we do not require user to call initialize before sending metrics. Then, if the user calls configuration routine, we change the state of the object as needed. We would need at least a way to enable aggregation on an object that already exists, ..."

File: tests/unit/dogstatsd/test_metrics.py [hunk 9]
```diff
@@ -52,29 +52,22 @@ def test_gauge_metric_aggregate(self):
         self.assertEqual(g.timestamp, 1713804588)
 
     def test_new_set_metric(self):
-        s = SetMetric("test", "value1", ["tag1", "tag2"], 1, 1713804588)
+        s = SetMetric("test", "value1", ["tag1", "tag2"], 1)
         self.assertEqual(s.data, {"value1"})
         self.assertEqual(s.name, "test")
         self.assertEqual(s.tags, ["tag1", "tag2"])
         self.assertEqual(s.rate, 1)
-        self.assertEqual(s.timestamp, 1713804588)
```
[the remark this answers is not in the document]
Comment (andrewqian2001datadog) [reply tests/unit/dogstatsd/test_metrics.py:60]: Yeah set metric doesn't have timestamp, felt kind of misleading to instantiate it with one.

File: tests/unit/dogstatsd/test_statsd.py [hunk 10]
```diff
@@ -803,7 +803,7 @@ def func(arg1, arg2, kwarg1=1, kwarg2=1):
         self.assertEqual('distributed.test', name)
         self.assert_almost_equal(500, float(value), 100)
 
-    def test_timed(self):
```
[the remark this answers is not in the document]
Comment (andrewqian2001datadog) [reply tests/unit/dogstatsd/test_statsd.py:806]: Yeah mb, thanks for the catch

File: datadog/dogstatsd/base.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (gh123man) [reply datadog/dogstatsd/base.py:659]: Oh awesome. Ill give this another look 👍

File: datadog/dogstatsd/base.py (hunk 2 above)
Comment (gh123man) [reply datadog/dogstatsd/base.py]: Spoke offline. seems like the linter wasn't running on this file before.

**Review by gh123man — DISMISSED** [review]
LGTM
Nice work on this 🚀

**Review by rayz — DISMISSED** [review]
looks good

File: datadog/__init__.py [hunk 11]
```diff
@@ -128,8 +138,11 @@ def initialize(
     if statsd_constant_tags:
         statsd.constant_tags += statsd_constant_tags
 
+    if statsd_disable_aggregating:
+        statsd.disable_aggregation()
+    else:
+        statsd.enable_aggregation(statsd_aggregation_flush_interval)
```
Comment (skarimo) [inline datadog/__init__.py:144]: Do you need to make sure buffering is disabled here?

**Review by skarimo — APPROVED** [review]

File: datadog/__init__.py (hunk 11 above)
Comment (andrewqian2001datadog) [reply datadog/__init__.py:144]: I think it should already be handled [here](https://github.com/DataDog/datadogpy/blob/add-client-side-aggregation/datadog/dogstatsd/base.py#L549) since we wont have a flush thread for buffering if aggregation is enabled

**Review by skarimo — DISMISSED** [review]

**Review by rayz — DISMISSED** [review]

**Review by rayz — APPROVED** [review]
